### PR TITLE
Fixes 7980

### LINKED ISSUE #7980: Pattern templates: incorrect argument matching when converters are present
The pattern templates can be false matched to an expression if there are converters defined.
The ``sigmatch.argtypeMatches`` is used to match arguments in ``patterns.nim`` and it considers converters.

Test case:
```nim
{.hint[Pattern]: true.}

type
  MyFloat = object
    val: float

converter to_myfloat*(x: float): MyFloat {.inline.} =
  MyFloat(val: x)

proc `+`(x1, x2: MyFloat): MyFloat =
  MyFloat(val: x1.val + x2.val)

proc `*`(x1, x2: MyFloat): MyFloat =
    MyFloat(val: x1.val * x2.val)

template optMul{`*`(a, 2.0)}(a: MyFloat): MyFloat =
  a + a

func floatMyFloat(x: MyFloat): MyFloat =
  result = x * 2.0

func floatDouble(x: float): float =
  result = x * 2.0
 ```

Compilation fails with error
  ``term_rewrite.nim(23, 14) Error: type mismatch: got (MyFloat) but expected 'float'``

Pattern template has rewritten ``floatDouble`` while it should not.

### PR DESCRIPTION
**Pull request by cooldome**
fixes #7980 by temporary disabling converters while pattern templates are being matched

### CODE REVIEW

**Comment by Araq** [comment]
Nothing wrong with this (yet?) but I need to understand the cause of this bug myself.

**Comment by Araq** [comment]
Hmm, sounds to me we want ``isSubtype, isEqual, isGeneric`` but not ``isConvertible``. After all we don't want any type conversions in the matching process, not even int16 to int32?

**Comment by cooldome** [comment]
I think so as well, no conversion at this phase. Pattern templates are applied after conversions already done.
What would be the right way to achieve this, extra optional argument on  ``sigmatch.argtypeMatches`` ? better ideas?

**Comment by Araq** [comment]
> What would be the right way to achieve this, extra optional argument on sigmatch.argtypeMatches ? better ideas?

Sounds good to me.

**Comment by cooldome** [comment]
New approach is ready for review

File: compiler/patterns.nim [hunk 1]
```diff
@@ -77,7 +77,7 @@ proc checkTypes(c: PPatternContext, p: PSym, n: PNode): bool =
   if isNil(n.typ):
     result = p.typ.kind in {tyVoid, tyStmt}
   else:
-    result = sigmatch.argtypeMatches(c.c, p.typ, n.typ)
+    result = sigmatch.argtypeMatches(c.c, p.typ, n.typ, from_hlo = true)
```
Comment (Araq) [inline compiler/patterns.nim]: Nitpick: It should be named ``fromHlo``, the Nim compiler does not use underscores.

Comment (cooldome) [reply compiler/patterns.nim]: np, changed to camel case

**Pull request merged**